### ISSUE TITLE
[XLA:CLIENT] Why force using U32 indices type when converting torch gather op

### ISSUE THREAD

**Issue by Nullkooland** [openxla/xla#15654]
This commit: https://github.com/openxla/xla/commit/a1c04b81ea9e2e23a362fdf5f7677800d2042f8f intrudoces the code that forces `xla::TorchGather` to convert indices tensor with element type of `i64`  to `u32`:
https://github.com/openxla/xla/blob/a1c04b81ea9e2e23a362fdf5f7677800d2042f8f/xla/client/lib/slicing.cc#L148-L152

This causes the StableHLO IRs exported with `torch_xla` to have such pattern around `stablehlo.gather`:

<img width="1185" alt="gather_ui32_indices" src="https://github.com/user-attachments/assets/0f170623-bbbf-42f9-9614-514252a110e8">

However, when lowering `stablehlo.gather` to MLIR tensor dialect's `tensor.gather`, the `tensor.gather` op requires that its `indices` operand tensor has *signless* integer type, so `ui32` would cause error.

@blakehechtman Could you have a look? This commit looks like a HACK, is it possible to remove this?

**Comment by cheshire** [comment]
The commit is 5 years old, and thus unlikely to be reverted. 

> However, when lowering stablehlo.gather to MLIR tensor dialect's tensor.gather

This seems outside of the scope of OpenXLA, more like a stablehlo/tensor dialect interop issue?

**Comment by Nullkooland** [comment]
> The commit is 5 years old, and thus unlikely to be reverted.
> 
> > However, when lowering stablehlo.gather to MLIR tensor dialect's tensor.gather
> 
> This seems outside of the scope of OpenXLA, more like a stablehlo/tensor dialect interop issue?

I don't think this is a stablehlo-to-tensor dialect conversion issue, since `stablehlo.gather` is able to take `i64` indices. After removing this force `u32` indices conversion code in XLA, the same torch gather will be exported as:
<img width="1443" alt="stablehlo_gather_i64_indices" src="https://github.com/user-attachments/assets/dcdb6480-a17f-4c9d-8ace-5dcdf34d5948">

I don't see why do we need this `u32` HACK.
